ops_omgx/monitor/transaction-monitor/services/monitor.service.js

```diff
@@ -0,0 +1,418 @@
+          receiptData = await this.getCrossDomainMessageStatusL1(receiptData);
+        }
+        await this.databaseService.insertReceiptData(receiptData);
+      }
+
+      // update scannedLastBlock
+      this.scannedLastBlock = this.latestBlock;
+
+      this.transactionMonitorSQL = false;
+      await this.endDatabaseService();
+
+
+      this.logger.info(`Found block, receipt and transaction data. Sleeping ${this.transactionMonitorInterval} ms...`);
+    } else {
+      this.logger.info('No new block found.');
+    }
+
+    await this.sleep(this.transactionMonitorInterval);
+  }
+
+  async startCrossDomainMessageMonitor() {
+    // connect to MySQL
+    this.crossDomainMessageMonitorSQL = true;
+    await this.startDatabaseService();
+
+
+    this.logger.info('Searching cross domain messages...');
+    const crossDomainData = await this.databaseService.getCrossDomainData();
+
+    // counts the number of server request
+    let promiseCount = 0;
+
+    let checkWhitelist = this.checkTime(this.whitelist);
+    if(checkWhitelist) await this.getWhitelist();
+    let checkNonWhitelist = this.checkTime(this.nonWhitelist);
+    
+    if (crossDomainData.length) {
+      this.logger.info('Found cross domain message.');
+      const promisesBlock = [], promisesReceipt = [];
+      for (let hashes of crossDomainData) {
+        // limits the rate of server request
+        if((promiseCount % this.L2rateLimit) === 0){
+          this.logger.info(`${this.L2rateLimit} promises reached!`);
+          await Promise.all(promisesBlock);
+          await Promise.all(promisesReceipt);
+          if(promiseCount % this.L2sleepThresh === 0){
+            await this.sleep(1000);
+          }
+        }
+        const hash = hashes.hash;
+        const blockNumber = Number(hashes.blockNumber);
+        promisesReceipt.push(this.L2Provider.getTransactionReceipt(hash));
+        promisesBlock.push(this.L2Provider.getBlockWithTransactions(blockNumber));
+        promiseCount = promiseCount + 1;
+      }
+
+      promiseCount = 0;
+      const blocksData = await Promise.all(promisesBlock);
+      const receiptsData = await Promise.all(promisesReceipt);
+
+      for (let receiptData of receiptsData) {
+        receiptData = await this.getCrossDomainMessageStatusL2(receiptData, blocksData);
+        
+        if(!receiptData.crossDomainMessage) continue;
+
+        // if its time check cross domain message finalization
+        if(checkWhitelist && this.whitelist.includes(receiptData.from)){
+          this.logger.info(`Checking message from whitelist address: ${receiptData.from}`);
+          receiptData = await this.getCrossDomainMessageStatusL1(receiptData, blocksData);
+        }else if (checkNonWhitelist && !this.whitelist.includes(receiptData.from)){
+          this.logger.info(`Checking message from non-whitelist`);
+          receiptData = await this.getCrossDomainMessageStatusL1(receiptData, blocksData);
+        }
+
```

Comment:
The logic is wrong. receiptData is from L2 block, but the whitelist contains the L1 contracts. You can't search the receipt.from or receipt.to in the whitelist. You have to get msgHash from `this.getCrossDomainMessageStatusL2` and use `getL1TransactionReceipt(msg.hash)` to check if the message is sent to the L1 contracts in the whitelist.

You can add a new column called fastRelay to the database. If the fastRelay is `true`, then you check the L1 receipt every 5 mins. Otherwise, you can check the L1 receipt every 3 hours.